Swaps: Add custom token flow - search by address and get it imported to your wallet

wachunei opened the pull request:
**Description**

This PR allows a user to enter a custom token address and use it to Swap


**Issue**

Resolves #2641

wachunei committed:
Add custom token flow

wachunei committed:
Update ErrorMessage snapshot

wachunei committed:
Add token endpoint

wachunei committed:
Update copy

wachunei committed:
Merge branch 'develop' into feature/swaps-custom-token

wachunei committed:
Merge branch 'develop' into feature/swaps-custom-token

sethkfman commented on app/components/UI/Swaps/utils/useFetchTokenMetadata.js:
```diff
@@ -0,0 +1,64 @@
+	'56': 'https://bsc-api.metaswap.codefi.network'
+};
+
+function useFetchTokenMetadata(address, chainId) {
```
Is this something that we will eventually move into a common Token Controller?

wachunei replied:
Good Question, at least I know is going to move to an API with a single endpoint with params `address, chainId` in the near future, that's why I left it there.

That endpoint could be exposed in a controller too, not planned so far tho.

sethkfman replied:
That makes sense. I guess if this utility starts to grow with more functionality we should check-in with Mark and the extension team to see if it would makes sense to convert it to a controller.

sethkfman commented on app/components/UI/Swaps/components/TokenImportModal.js:
```diff
@@ -0,0 +1,115 @@
+	}
+});
+
+function TokenImportModal({ isVisible, dismiss, token, onPressImport }) {
```
Is there any risk that `token` could be undefined?

wachunei replied:
No, This in only rendered with valid metadata https://github.com/MetaMask/metamask-mobile/pull/2729/files/0aac78f607623c019b480d0f8e2bfa61494859a1#diff-8bdf03ea62d01afd27f35a8881d23e3175b5a9797c43372a7e14f117612bb133R348

wachunei committed:
Allow custom tokens in the flow

wachunei committed:
Merge branch 'develop' into feature/swaps-custom-token

sethkfman reviewed (approved):
LGTM

sethkfman committed:
Merge branch 'develop' into feature/swaps-custom-token

wachunei committed:
Add import analytics

wachunei committed:
Merge branch 'feature/swaps-custom-token' of github.com:MetaMask/metamask-mobile into feature/swaps-custom-token

wachunei committed:
Merge branch 'develop' into feature/swaps-custom-token

wachunei committed:
Merge branch 'develop' into feature/swaps-custom-token

mobularay commented:
@jakehaugen: I just added the "design-qa" label

wachunei committed:
Merge branch 'develop' into feature/swaps-custom-token

wachunei committed:
Fix navigation old code

ibrahimtaveras00 reviewed (changes requested):
**Issue 1:** 

it’s a bit tight to scroll on small device to see the full “can’t find a token” message box
http://recordit.co/Qd3Qsgce0D

**Issue 2:**

when you tap on the bscscan link it takes you to a 403 = http://recordit.co/zb2ZWCDbKK

I searched this token = 0x170dca460d661e834ff91bf6011afa8aa1a00b05

wachunei committed:
Fix scroll and url

jakehaugen commented:
> **Issue 1:**
> 
> it’s a bit tight to scroll on small device to see the full “can’t find a token” message box
> http://recordit.co/Qd3Qsgce0D

No idea why this would cause a scrolling issue. Are design changes needed?

wachunei commented:
@jakehaugen I fixed it, no design changes needed. The issue happened because the footer was a regular view that did not capture the scrolling gesture. I changed it to a Touchable (without feedback) and behaves correctly now.

ibrahimtaveras00 reviewed (approved):
Fixes look good, QA Passed 👍🏽

wachunei committed:
Merge branch 'develop' into feature/swaps-custom-token
